rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,165 @@
+
+A positive point for having standalone mappers is that it's language agnostic: mapper developers can write the logic in whatever language they want, and it's fine, until they follow the guidelines for the Engine -> Mapper communication protocol.
+
+One negative point is that they'll be provided with the AWS credentials, and they can perform out-of-band data extraction on it on both of the above ways. The security profile for this approach is the worst of all.
+
+The other issue with this way is that it'll be at least as hard as the ways under [Mapper Logic Inside Postgres](#Mapper-Logic-Inside-Postgres) section. So we need a "virtual" mapper to be created and the calls to that mapper should be funneled into these standalone mappers. Also, at the module level we need the entity creation logic (and uninstall logic, of course) to somehow be transmitted to the engine. Given all these, together with the security issues that are purely "wild west situation", I don't think if this would be a good approach neither in the short, nor in long term.
+
+## In-IaSQL-Container Mappers
+
+In this approach, the logic of the mappers will be executed on IaSQL: either on the Postgres server or on the Engine. Our current mappers are under this category.
+- Mapper logic is in IaSQL engine: `aws_acm`'s `CertificateMapper`, `aws_ecr`'s `RepositoryImageMapper`, etc.
+- Mapper logic is in Postgres server: This is where we initially called Pure-SQL module. Note that `aws_ecs_simplified` module can't be considered under this category, since it introduces no new mappers. It's playing in module-level (not mapper level) and handling a new Entity using already-existing mappers.
+
+Before diving into "Pure-SQL Modules" category, first let's see if our current mappers are surjective to the space of "Mapper logics inside IaSQL engine".
+
+### Mapper Logic Inside IaSQL Engine
+
+Currently, we're developing our mappers in this category. But there might be other ways to integrate the logic inside the engine, without having to exactly follow our current pattern. I'm getting inspired by [Strapi](https://github.com/strapi/strapi), which is a headless CMS. You can create models using a GUI from Strapi's admin panel and the database will be changed on-air, new API routes will be added with proper logic, and it achieves these without losing performance or being in need of pushing new files to the codebase (but the codebase is changing on-air).
+
+So we can get inspired and create on-air mapper codes. But to ensure safety, we might take one of the two approaches: either by narrowing down the available variables, or the by narrowing down the available functionality.
+- Create a sandbox and run mapper's JavaScript code on IaSQL engine: in that sandbox, we can narrow down the context and not pass variables that might cause problems.
```

Comment:
This option should be pulled out and turned into the Proposal with all of the rest put into Alternatives Considered.

Reply:
It would also be good to bring up the SDK vs external module install are orthogonal.

We can break this into a few different tasks:
- Creating the VM2-based code runner and RPC layer, which can be validated with a handwritten internal module to flesh it out, perhaps simply porting a smaller existing module, like the S3 module, and renaming it to `aws_s3_vm2` for this first piece.
- Creating the SDK that generates the bundle for VM2 to run, re-porting the chosen test module again, and seeing how close it looks to the original module.
- Modifying `iasql_install` to handle installing modules from a URL, where the URL is pointing to the bundled `.js` file that the SDK can generate (perhaps with some sort of signing mechanism to be more sure it was built via the SDK?) and the relevant changes necessary to deal with dependency declaration and management in that world.

Technically each one depends on the one before it, but a lot of the foundational work of those three main tasks can be done in parallel, but finishing it off requires the first to be done first, second done second, etc.

Reply:
> * Creating the VM2-based code runner and RPC layer, which can be validated with a handwritten internal module to flesh it out, perhaps simply porting a smaller existing module, like the S3 module, and renaming it to `aws_s3_vm2` for this first piece.

do we expect any performance penalization running the code in this isolated environment?

Reply:
Extracting the modules from the engine technically would make easier the testing since when any change is done in a module does not need to wait for all other modules, but this raises a couple of questions:

- what happens if a module depends on another one? If the other module change, the dependent module would be affected afterwards? would we need to introduce some versioning?
- Any change done in the engine should trigger integration tests in all modules? or again we would need to add some versioning to the engine too?

Reply:
> do we expect any performance penalization running the code in this isolated environment?

A minor amount that should not be measurable vs the CPU-bound XML parsing and TypeORM object-to-query/query-to-object transforms and the IO-bound network requests to AWS.

The last time I checked our own code was less than 1% of the CPU time and far less than 1% of the total time, depending on the operation being performed (eg, the RDS tests taking half an hour, but code we wrote taking up approximately 1 second of that time).

Reply:
> Extracting the modules from the engine technically would make easier the testing since when any change is done in a module does not need to wait for all other modules, but this raises a couple of questions:
> 
>     * what happens if a module depends on another one? If the other module change, the dependent module would be affected afterwards? would we need to introduce some versioning?

We still have the versioning in the database to support modules with version X depending on other modules with version Y in the `iasql_module` table. The majority of work would be tweaks to the dashboard to understand module versions better.

As far as how testing of modules outside of the repo would work, the SDK would have to require something like the `module.json` but with more precision on the dependency list, probably closer to Node's `package.json`? But we'd probably hew closer to a lockfile and not let modules float their dependencies, even if that causes more versioning churn on "high level" dependencies, to reduce risks with bad module interactions.

Then testing wouldn't really be any different: the module test installs the dependencies that it needs and confirms everything still works.

>     * Any change done in the engine should trigger integration tests in all modules? or again we would need to add some versioning to the engine too?

This is why I made it an error for a module to not depend on `iasql_platform`, as this version will basically become the engine version in the future, and external modules need to re-publish to function on a newer version of the engine.

In a true ecosystem this would slow down adoption of the newest version of the engine, but in this world updates to the engine itself would be shorter (and once we're actually on semver we could perhaps make it so that you can only pin to `X.Y` and not `X.Y.Z` so bugfix releases aren't blocked on module authors to re-target).

Reply:
Ok, I got it. That sounds great to me then. I'd suggest including a bit about the versioning in the final proposal, and maybe talk a bit more on what we expect to expose in the SDK without going into implementation details.